lib/upipe-dvb/upipe_dvbsrc.c

```diff
@@ -0,0 +1,688 @@
+    struct uchain request_list;
+
+    /** upump manager */
+    struct upump_mgr *upump_mgr;
+    /** read watcher */
+    struct upump *upump;
+
+    /** DVB receiver uri */
+    char *uri;
+
+    /** DVB device */
+    struct dvb_device *dvb;
+
+    /** DVB demux */
+    struct dvb_open_descriptor *demux;
+
+    /** DVB frontend */
+    struct dvb_open_descriptor *frontend;
+
+    /** public upipe structure */
+    struct upipe upipe;
+};
+
+UPIPE_HELPER_UPIPE(upipe_dvbsrc, upipe, UPIPE_DVBSRC_SIGNATURE)
+UPIPE_HELPER_UREFCOUNT(upipe_dvbsrc, urefcount, upipe_dvbsrc_free)
+UPIPE_HELPER_VOID(upipe_dvbsrc)
+
+UPIPE_HELPER_OUTPUT(upipe_dvbsrc, output, flow_def, output_state, request_list)
+UPIPE_HELPER_UREF_MGR(upipe_dvbsrc, uref_mgr, uref_mgr_request,
+                      upipe_dvbsrc_check,
+                      upipe_dvbsrc_register_output_request,
+                      upipe_dvbsrc_unregister_output_request)
+UPIPE_HELPER_UBUF_MGR(upipe_dvbsrc, ubuf_mgr, flow_format, ubuf_mgr_request,
+                      upipe_dvbsrc_check,
+                      upipe_dvbsrc_register_output_request,
+                      upipe_dvbsrc_unregister_output_request)
+UPIPE_HELPER_UCLOCK(upipe_dvbsrc, uclock, uclock_request, upipe_dvbsrc_check,
+                    upipe_dvbsrc_register_output_request,
+                    upipe_dvbsrc_unregister_output_request)
+
+UPIPE_HELPER_UPUMP_MGR(upipe_dvbsrc, upump_mgr)
+UPIPE_HELPER_UPUMP(upipe_dvbsrc, upump, upump_mgr)
+
+static void upipe_dvbsrc_log(void *priv, int level, const char *fmt,...)
+{
+    int loglevel;
+    switch (level) {
+    case LOG_WARNING:
+    case LOG_ERR:
+    default:
+        loglevel = UPROBE_LOG_WARNING;
+        break;
+    case LOG_NOTICE:
+        loglevel = UPROBE_LOG_NOTICE;
+        break;
+    case LOG_DEBUG:
+    case LOG_INFO:
+        loglevel = UPROBE_LOG_DEBUG;
+        break;
+        break;
+    }
+
+    struct uprobe *uprobe = priv;
+
+    va_list args;
+    va_start(args, fmt);
+
+    size_t len = vsnprintf(NULL, 0, fmt, args);
+    va_end(args);
+    if (len > 0) {
+        char str[len + 1];
+        va_start(args, fmt);
+        vsnprintf(str, len + 1, fmt, args);
+        va_end(args);
+        char *end = str + len - 1;
+        if (isspace(*end)) {
+            *end = '\0';
+        }
+        uprobe_log(uprobe, NULL, loglevel, str);
+    }
+}
+
+/** @internal @This allocates a DVB receiver source pipe.
+ *
+ * @param mgr common management structure
+ * @param uprobe structure used to raise events
+ * @param signature signature of the pipe allocator
+ * @param args optional arguments
+ * @return pointer to upipe or NULL in case of allocation error
+ */
+static struct upipe *upipe_dvbsrc_alloc(struct upipe_mgr *mgr,
+                                        struct uprobe *uprobe,
+                                        uint32_t signature, va_list args)
+{
+    struct upipe *upipe = upipe_dvbsrc_alloc_void(mgr, uprobe, signature, args);
```

Comment:
You have to check for NULL. Some pipes may try to allocate this pipe with an incompatible signature (auto inner pipe for instance).